tlair/src/main/scala/at/forsyte/apalache/tla/lir/TypeLevelBuilder.scala

```diff
@@ -0,0 +1,186 @@
+package at.forsyte.apalache.tla.lir
+
+import at.forsyte.apalache.tla.lir.oper.{TlaArithOper, TlaBoolOper, TlaFunOper, TlaOper}
+
+// We use shapeless for facilitating the type-level encodings of TlaExpressions
+import shapeless._
+import poly._
+
+// A (very) rough sketch at what TLA Expressions might look like when
+// we represent out typing discipline inside of Scala's type system.
+object Tla {
+  // The type of any TLA expression
+  sealed trait Exp[T] {
+    def toString: String
+  }
+
+  // Just an alias
+  type IntT = BigInt
+
+  // A value constructs a TLA expression tagged with the Scala type it carries as a payload.
+  class Value[T](value: T) extends Exp[T]
+  case class Int(value: IntT) extends Value(value)
+  case class Str(value: String) extends Value(value)
+  case class Bool(value: Boolean) extends Value(value)
+
+  // Each compound expression is reprsented by a new type
+
+  sealed trait SeqType[T]
+  case class TlaSeq[T](seq: Seq[Exp[T]]) extends Exp[SeqType[T]]
+
+  sealed trait TupleType[A, B]
+  case class Empty() extends Exp[TupleType[Unit, Unit]]
+  case class Single[A](fst: A) extends Exp[TupleType[A, Unit]]
+  case class Tuple[A, B](fst: A, snd:B) extends Exp[TupleType[A, B]]
+
+  sealed trait RecType[Fields]
+  case class Rec[Fields <: HList](fields: Fields) extends Exp[RecType[Fields]]
+
+  sealed trait OperType[Args <: HList, Ret]
+
+  // Names are expressions
+  // The idea here is that names also record the type of the expression
+  // they refer to. When we don't know what that type is, we might use
+  // `Name[Exp[Any]]`.
+  case class Name[T](name: String) extends Exp[T]
+
+  // An unapplied operator
+  case class Op[Args <: HList, Ret](oper: TlaOper)
+      extends Exp[OperType[Args, Ret]]
+
+  // Application of an operator
+  case class App[Args <: HList, Ret](op: Exp[OperType[Args, Ret]], args: Args)
+      extends Exp[Ret]
+
+  // Binary operators of type (T, T) => Ret
+  type BinOpApp[T, Ret] = App[T :: T :: HNil, Ret]
+  type Tup[A, B] = Exp[TupleType[A, B]]
+}
+
+// Utility functions for constructing expressions. Most of these don't actually do much,
+// but a few are useful.
+object TypeLevelBuilder {
+  import Tla._
+
+  def binOp[T, Ret](op: TlaOper)(a: T, b: T): Exp[Ret] = App(Op(op), a :: b :: HNil)
+
+  def plus: (Exp[IntT], Exp[IntT]) => Exp[Tla.IntT] =
+    binOp(TlaArithOper.plus)
+
+  // A polymorphic operator
+  def equal[T]: (Exp[T], Exp[T]) => Exp[Boolean] =
+    binOp(TlaOper.eq)
+
+  def and(a: Exp[Boolean], b: Exp[Boolean]): Exp[Boolean] =
+    binOp(TlaBoolOper.and)(a, b)
+
+  // Variadic conjunction
+  def and(args: Exp[Boolean]*): Exp[Boolean] = args match {
```

Comment:
Variadic conjunctions are quite common in TLA+. So it would be good to have it as a normal form, instead of unwinding them into binary conjunctions.

Reply:
I think that should be pretty feasible! I'd be interested to learn the stakes here. On the face of it, I'd assume variadic conjunction were just syntax suggar for chained binary conjunctions, but does it affect performance or the encoding to SMT somehow?

Reply:
It's just harder to write some transformations when you have hanging trees of and-s and or-s. I would suspect that SMT solvers group conjunctions and disjunctions into lists during the preprocessing stage. But that might be harder to do, since our encoding would produce multiple constraints for non-variadic conjunctions and disjunctions.